be2-scala/src/test/scala/ch/epfl/pop/decentralized/GossipManagerSuite.scala

```diff
@@ -0,0 +1,135 @@
+
+  private var inMemoryStorage: InMemoryStorage = _
+  private var messageRegistry: MessageRegistry = _
+  private var pubSubMediatorRef: ActorRef = _
+  private var dbActorRef: AskableActorRef = _
+  private var securityModuleActorRef: AskableActorRef = _
+  private var monitorRef: ActorRef = _
+  private var connectionMediatorRef: AskableActorRef = _
+
+  override def beforeEach(): Unit = {
+    inMemoryStorage = InMemoryStorage()
+    messageRegistry = MessageRegistry()
+    pubSubMediatorRef = system.actorOf(PubSubMediator.props)
+    dbActorRef = system.actorOf(Props(DbActor(pubSubMediatorRef, messageRegistry, inMemoryStorage)))
+    securityModuleActorRef = system.actorOf(Props(SecurityModuleActor(RuntimeEnvironment.securityPath)))
+    monitorRef = system.actorOf(Monitor.props(dbActorRef))
+    connectionMediatorRef = system.actorOf(ConnectionMediator.props(monitorRef, pubSubMediatorRef, dbActorRef, securityModuleActorRef, messageRegistry))
+
+  }
+
+  val pathCorrectRumor: String = "src/test/scala/util/examples/json/rumor/rumor.json"
+
+  val rumorRequest: JsonRpcRequest = JsonRpcRequest.buildFromJson(readJsonFromPath(pathCorrectRumor))
+
+  val rumor: Rumor = rumorRequest.getParams.asInstanceOf[Rumor]
+
+  test("gossip handler should forward a rumor to a random server") {
+
+    val gossipManager: ActorRef = system.actorOf(GossipManager.props(dbActorRef, monitorRef, connectionMediatorRef))
+    val gossipHandler = GossipManager.gossipHandler(gossipManager)
+
+    val peerServer = TestProbe()
+
+    // register server
+    connectionMediatorRef ? ConnectionMediator.NewServerConnected(peerServer.ref, GreetServer(PublicKey(Base64Data("")), "", ""))
+
+    val output = Source.single(Right(rumorRequest)).via(gossipHandler).runWith(Sink.head)
+
+    Await.result(output, duration)
+
+    peerServer.expectMsg(duration, ClientAnswer(Right(rumorRequest)))
+  }
+
+  test("gossip handler should send to only one server if multiples are present") {
+    val gossipManager: ActorRef = system.actorOf(GossipManager.props(dbActorRef, monitorRef, connectionMediatorRef))
+    val gossipHandler = GossipManager.gossipHandler(gossipManager)
+
+    val peerServer1 = TestProbe()
+    val peerServer2 = TestProbe()
+    val peerServer3 = TestProbe()
+    val peerServer4 = TestProbe()
+
+    val peers = List(peerServer1, peerServer2, peerServer3, peerServer4)
+
+    // register server
+    for (peer <- peers) {
+      connectionMediatorRef ? ConnectionMediator.NewServerConnected(peer.ref, GreetServer(PublicKey(Base64Data("")), "", ""))
+    }
+
+    val output = Source.single(Right(rumorRequest)).via(gossipHandler).runWith(Sink.head)
+
+    Await.result(output, duration)
+
+    peers.map(_.receiveOne(duration)).count(_ != null) shouldBe 1
+
+  }
+
+  test("gossip handler should send rumor if there is an ongoing gossip protocol") {
```

Comment:
Not sure I understand what you are testing here ?  How is it different from the previous test ?

Reply:
It asserts that the gossip protocol goes on as expected (do not send to a peer that already received). But looking back it is true that it includes what previous tests is doing. Previous test can be removed imo.

Reply:
Okay, I got it now, you can keep the previous one, it will be easier to spot the break in it if that part fails. This last test is quite difficult to understand, I would love to see some comments for the major steps.